nav2_behavior_tree/include/nav2_behavior_tree/bt_action_node.hpp

```diff
@@ -174,7 +179,24 @@ class BtActionNode : public BT::ActionNodeBase
       // user defined callback
       on_tick();
 
-      on_new_goal_received();
+      send_new_goal();
+    }
+
+    // if new goal was sent and action server has not yet responded
+    // check the future goal handle
+    if (goal_sent_) {
+      if (!check_future_goal_handle()) {
```

Comment:
This logic will bypass on_wait_for_result which in turn will stop checks for path preemption in the FollowPath BT node. Is this block necessary if just below you have the same thing in L213-223?

Reply:
The condition on L208 is true when there's a new goal that has to be sent to the action server while a previous goal is being executed. So L213-223 is used only to immediately check the future goal handle after a new goal has been sent **while a previous goal was being executed**. If `spin_until_future_complete` times out at this iteration, in the next iteration `goal_updated_` would be false and `check_future_goal_handle()` would never be called again. The L187-198 block is the main block that checks the future goal handle at every iteration if the future has not completed yet.

It is my understanding that `on_wait_for_result` is only meant to run when the BT node is waiting for the accepted action to complete, not when the BT node is waiting for the action server to respond to a `send_goal`. Is that right? In that case, `on_wait_for_result` should be bypassed until the future goal handle completes.

Reply:
> It is my understanding that on_wait_for_result is only meant to run when the BT node is waiting for the accepted action to complete, not when the BT node is waiting for the action server to respond to a send_goal. 

Thinking about this from a client-server perspective, I don't think the client (the BT node) cares about what the state of its request is in (sending goal, processing goal, etc) since for all it cares, once it sends, its being crunched. So I'd assume that would be called too any time its processing. Though knowing what FollowPath does, I'm not sure if having a new goal matters when its still sending the current goal - I don't think we can preempt that easily so maybe you're right we should keep it separate. 

I'm concerned about the `goal_sent_` variable - its not reset on failure / halt / terminal conditions. I think it could cause state to bleed into multiple requests. For instance, its only set back to false once accepted, but if rejected, it remains true which would impact the next request

Reply:
Yup, that's right. I've missed a few important places where `goal_sent_` should have been reset.